### PR TITLE
Runechat - stop bugging me for gods sake

### PR DESCRIPTION
**Pull request by AffectedArc07**
# NOTE: This PR has been entirely re-written with new screenshots and a rewrite of the description

## What Does This PR Do
image

This PR adds runechat. It is not the same as TGs implementation (specifically we dont have popups for radio or emotes), and thats becaus at 150 pop the server cant take it. Im not sacrificing performance for a feature as minor as this.

You will only see a runechat above someones head if you would actually see a message in chat. Examples include:
- You wont see a whisper runechat if you cannot hear that person perfectly (you will see the asterisked out message though)
- You wont see runechats from people speaking down radio channels you cant hear
- You wont see runechats on people out of range
- You will not see a message in a specified language if you do not have said language

If you are a crewmember, your message colour is picked from either your hair colour, your body colour if your species lets you set body colour, or your eye colour if you are a grey. Screenshots below
image
image
image
image

Language colouring takes precedence over anything else
image
image

Credit to @blinkdog for the HTML strip-out code used in parts of this. Everything else is ported from TG, with the exception of the radio icon colouring and species stuff colouring, which I did myself.

# Again, I must state that this PR **will not** be merged if the performance is not up to par. Yes its a cool feature, but server performance is a top priority, and that is not changing

## Why It's Good For The Game
1: People like this 
2: I am fed up of people saying "AA WHEN ARE YOU PORTING RUNECHAT"

## Changelog
:cl: AffectedArc07
add: Runechat
/:cl:

### CODE REVIEW

**Comment by ppi13** [comment]
This is how god intended ss13 to be played.

**Comment by TDSSS** [comment]
Is this a toggle? Please tell me I can just ignore this if it's horrible.

**Comment by datlo** [comment]
This definitely should be a toggle if it's added at all which I am not sure it should.

**Comment by ZomgPonies** [comment]
Probably want to keep colors strictly for languages, like it is in textual chat. You can use different font sizes for whispers/emotes (Fulp has a really clean runechat emote implementation). Don't forget stuff like loudspeaker, voice of god, etc.

**Comment by AffectedArc07** [comment]
Yes, this is a toggle, dont worry.

@ZomgPonies this is a barebones PR for TMing only. If the heads and maints deciede its good, ill polish it and add all the other pizzazz in.

**Comment by Aurorablade** [comment]
Its horrible and a jumbled mess by the pictures and I do not feel we need this.

**Comment by ZomgPonies** [comment]
It's nowhere as bad as the pictures -- Spend a round observing on any other popular server to get an actually accurate feel for how busy it is on screen.

**Comment by Aurorablade** [comment]
Runechat, ultima online chat call it what you will I have been around many games with this type of thing in a isometric envrioment.it sucks. It's more tolerable in a 3d environment

**Comment by AffectedArc07** [comment]
image
image
image

image
@ZomgPonies  Considering we have anywhere from 2-4x the population of TG, not to mention more RP and a smaller station, I can see this going horribly wrong.

TG doesnt *have* prisoner processing. This is their processing area on meta
image
If the grime on the floor doesnt give it away, its never used because people are brigged into the cells directly

Here is an average sized runechat message being simulated over our prisoner processing area, fully loaded (which can and has happened more often that not)
image

You're telling me this is readable?

**Comment by Tokorizo** [comment]
LETS FUCKING GOOOOOOOOOOOOOOOOO

**Comment by Carthusia** [comment]
I look forward to this, so long as its a toggle, the people who would love this are happy, and the people who dont, are happy.

I like the idea of being able to focus on one screen rather than having to teach my brain to seperate my eyes in seperate directions like im an inguana to see both the main action screen, and a text screen like a 1980's Text RPG

**Comment by AffectedArc07** [comment]
Putting this in a comment to make it more noticable from the PR description. The point of this testmerge is not just a "do we want this", theres a large performance factor at play here too. In particular, 2 things will be measured.

1: General profile of the server (Whats taking longest to execute). The messages heavily rely on calls to `animate()`, which is one of the most expensive operations in the entire game
2: GC rate of messages (In technical terms, when these things delete do they still have references being held)

This is in no way a final product, nor do I have intention to actually get this fully merged in, given what I know about BYOND and how badly this system will handle at 120 population.

**Comment by AffectedArc07** [comment]
> LETS FUCKING GOOOOOOOOOOOOOOOOO
> also, you people do realize you can TURN IT OFF if you don't wanna see it right? like cmon, this argument isn't logical because there is an option to DISABLE it. everyone acting like there being forced to play with it, as a person whos wanted this for a long time I DESPISE having to read the textbox and move my eyes to the game just to see it's flooded with chaos and other shit, this is a quality of change everyone should be on board about because not only is it UI friendly, it also appeases both sides, the people who DO and DONT want it. And @AffectedArc07, the poison well tactic, and garbage example from Reddit are pretty funny and all but you really trying to make it seem like its a bad PR when almost all other servers have this in-game, there's a reason why it's so popular. Oh and if you really need to find a specific part of the text you missed, the text box is STILL there.

@Tokorizo lets break this down, piece by piece.

> also, you people do realize you can TURN IT OFF if you don't wanna see it right? like cmon, this argument isn't logical because there is an option to DISABLE it. everyone acting like there being forced to play with it

Yes it can be disabled, but you shouldnt be expected to have to modify your preferences every time you move in and out of a highly populated area, thats bad game design. No one is acting like they are being "forced to play with it". I mentioned in a comment yesterday that it will be a toggle, and have since updated the PR description to make it even more obvious that its a toggle. No one is being forced here.


> And @AffectedArc07, the poison well tactic, and garbage example from Reddit are pretty funny and all

This screenshot has been posted around multiple places, and is the prime example of why runechat does not work in crowded environments. I also dont agree with calling this "poisoning the well". The entire point of this PR, is to show it to the server population, with examples of why it wont work here. The whole point is for it to be testmerged for player satisfaction, and for performance (Keep in mind, this puts substantial load on SStimer, and relies a lot on `animate()` calls


>  almost all other servers have this in-game, there's a reason why it's so popular.

I said this before, and I will say it again. It works on servers like TG because they dont have an RP level. The processing situation literally cannot exist on TG, because this is their processing area, on DeltaStation keep in mind, their biggest map.
image
TG doesnt have the same situations we do. A lot of HRP servers still dont have this because of the issues it creates (Take Bay12 and AuroraStation for example). It does not work in high RP environments where you will have multiple people in a small area.

**Comment by Krysonism** [comment]
Neither the screenshot in the OP or the mock-up is representative of how runechat actually looks in-game.

I am pretty sure the screenshot in OP was taken while the timer subsystem broke and the messages were not clearing properly.

In my opinion runechat facilities RP by making it easier notice when someone is trying to talk to you.

**Comment by procdrone** [comment]
I mean **Affected**, staff and you did this to yourselves by going super hush hush on why you wouldn't add it in the first place, refusing to give any explanation.

Other than that, a very welcome addition.

Also don't say we have TONS more roleplay. server been edging into LRP side for quite some time now. but, thats not a place to discuss state of the server.

Additionally, you show TG shuttle as an example why bad. That shuttle is tiny. and its easier to notice someone saying something by this than constantly FLOWING textbox. 

Also sure we have 120 people. 120 throughout the whole station. rarely people pile up in one place except shuttle.

Sure, there might be downsides but there is a reason why almost everyone adopted this.

**Comment by Carthusia** [comment]
Something I have to note as well after last nights shuttle, but the picture of chat all over the place looking really bad, is more or less the same as what we have now in high populated areas where it's nearly impossible to read what's being said to you during lots of action.

I find on a lot of shuttles I cant really even talk anymore due to the high amount of red text actions or other chat being done.

Though other than extreme high pop areas like that, I feel like this would be more or less fine to witness, I look forward to its test to see for myself though.

**Comment by Dumbdumn5** [comment]
_Personally?_ I think runechat is cool as fuck. I don't know the technical bit about animate calls though. Do you have any way you could see how much it'd impact CPU load for those of us not into the wizard speak but who think it's a neat idea regardless of the potential visual clutter? Can have it default to off and just let those who like it turn it on for that particular problem.

**Comment by Tokorizo** [comment]
@AffectedArc07 
Allow me to return the favor, and thanks for being so calm and respectful in your rebuttal.

> Yes it can be disabled, but you shouldnt be expected to have to modify your preferences every time you move in and out of a highly populated area, thats bad game design. No one is acting like they are being "forced to play with it". I mentioned in a comment yesterday that it will be a toggle, and have since updated the PR description to make it even more obvious that its a toggle. No one is being forced here.

Firstly I'm sure a majority of people will always keep rune chat on, even if the situation is chaotic. I doubt every time the situation is chaotic people will turn it off and that shouldn't be the mentality for all players. Its MUCH better than attempting to read the chatbox with emotes, actions, comms and other minor things that flood the chat and make it INCREDIBLY difficult to read what is happening, it's so constant that people will usually miss people talking to them. This problem is NOT just "Runechat" its the chaotic nature of SS13 and its WORST when I'm forced to try and have a conversation with someone on the shuttle only for what I've said previously to be spamming the chat. Its WORST gameplay design to have your eyes moving back and forth between the chatbox and gameplay window while also having said chatbox moving at the speed of light due to spam from other sources. I repeat, a quality of life change that keeps the important stuff on one screen, dialogue and gameplay. So no, I disagree that its "bad" gameplay design compared to what we currently have.

> This screenshot has been posted around multiple places, and is the prime example of why runechat does not work in crowded environments. I also dont agree with calling this "poisoning the well". The entire point of this PR, is to show it to the server population, with examples of why it wont work here. The whole point is for it to be testmerged for player satisfaction, and for performance (Keep in mind, this puts substantial load on SStimer, and relies a lot on `animate()` calls

I don't think this is the fault of runechat for being such a chaotic scene. Imagine if I was trying to keep track of this in the CHATBOX, crowded environments are ALWAYS chaotic, I get a headache every time there are 6-sec officers in processing and they're spamming a cup of water in the chatbox making me miss things. Also what @Krysonism said...

> Neither the screenshot in the OP or the mock-up is representative of how runechat actually looks in-game.
> 
> I am pretty sure the screenshot in OP was taken while the timer subsystem broke and the messages were not clearing properly.

I more inclined to believe that this was a bug with the system that resulted in messages not clearing because that would mean ALL players had to have said something at the same time which really isn't the common case for runechat. I occasionally play fulp (very similar to /tg/ but better) and I've never had a headache from reading runechat text, and ive never seen the example you've used actually happen. The picture you use is a bad example not only because I'm certain this picture is a bug to show how "bad" this feature is, but you also leave out any interesting and appealing things in the first page, and yes I understand YOU don't like it, but this goes back to what I was saying about poisoning the well. Also the "it won't work here" isn't really fair to say, its a common PR that gets used on server pops higher than ours (CM, Fulp, Etc) so if this really didn't work on other servers I'm a bit surprised why they haven't just removed it... maybe because its a popular PR that many people agree is a major quality of life change. Also, I understand the point behind optimization and that's a fair and valid point against rune chat no question about it. But I'm more inclined to believe we'll be fine, if CM and Fulp which sometimes have over 120+ can handle it, I'm not sure why we can't. But again well see what happens. 

> I said this before, and I will say it again. It works on servers like TG because they don't have an RP level. The processing situation literally cannot exist on TG, because this is their processing area, on DeltaStation keep in mind, their biggest map.
> image
> TG doesnt have the same situations we do. A lot of HRP servers still dont have this because of the issues it creates (Take Bay12 and AuroraStation for example). It does not work in high RP environments where you will have multiple people in a small area.

I feel like a more fair comparison for RP levels and an example for this PR would be fulpstation, their MRP, have a similar player count than us, and actually have a lot of the features we want. I also don't really understand what the processing example is, sorry I don't seem to understand. Also, I think the reason why Bay and Aurora don't have this feature is that at most their player count is 50, so they really don't have chaotic moments. And what do you mean "multiple people in a small area" are we talking three people in the bridge? 6 people in processing? 50 people on the shuttle? I think it can work with multiple people in a small area, that's the issue with having such a high player count, it's going to be CHAOTIC, rune chat or not.

But again, I respectfully disagree with most of your points but I do agree on a possible performance issue, I'm sure there is something we can do about it and again, looking forward to this PR.

**Comment by shazbot194** [comment]
> [https://user-images.githubusercontent.com/25063394/91082783-1488e500-e641-11ea-8d43-56b8d53abd4f.png](img)

I don't think this is at all readable, and this is only roughly 14 people talking, let along 20, 30, or 100. This would be awful for any sizable group of people more than 3 people.

**Comment by Tokorizo** [comment]
> image
> 
> I don't think this is at all readable, and this is only roughly 14 people talking, let along 20, 30, or 100. This would be awful for any sizable group of people more than 3 people.

This picture is more than likely a bug and doesn't show a good example of what it would look like in-game, this is a poor extreme compared to how it actually would look.

**Comment by Tarhalindur** [comment]
Test merge seems to indicate that Rune is broken with tcomms, it says "says" twice.

**Comment by Tarhalindur** [comment]
Also, there's another fact about runechat. What does this actually add?

It doesn't let you ignore normal chat, because radio still appears there so you need to look there anyway. So I don't get what it adds beyond random graphical fluff that we don't really need and that is potentially detrimental to the server.

**Comment by AffectedArc07** [comment]
@Tarhalindur fixed, spotted this during testing but never pushed the commit since it was on another device.

As for what it actually adds, youre completely right. It is just graphical fluff, you still have to look at chat for radio and other things. I really do not see what all the hype over this system is.

**Comment by Tarhalindur** [comment]
Honestly it might be worth holding a thread on the forums or something asking what the hype is about. There sure doesn't seem to be a convenience advantage as already stated.
The only real rationale I can think of is it's just people going "COOL TEXT WEW", but I don't think that's a good reason to merge something. *Especially* a performance-costly PR.

**Comment by procdrone** [comment]
@Tarhalindur  If you don't like it you can always toggle it off. 

Performance issues are valid though. So, for actual performance. @AffectedArc07 what did the testmerge tell us?

**Comment by AffectedArc07** [comment]
After the testmerge, the following conclusions have been drawn.
Design wise, 9 people out of 150 disabled the preference, and the clumping issues weren't as bad as expected. I was unable to test the shuttle scenario due to the fact an admin spawned in a micro shuttle which completely muddied the test data. *sigh*

Now onto the bad side, performance. 
```
/datum/chatmessage
qdel() Count: 367958
Destroy() Cost: 9018ms
```
These statistics were observed in a 1 hour 30 minute round. 367,958 chat messages were created (One per client HUD, hence why this option is even toggleable), and took 9018ms to destroy them all **in total**, roughly 0.024ms per delete. The issue here, is the sheer volume of messages.
image

The sheer volume of messages caused the garbage collector to have a queue of **6483** things to delete. Thats not ok at all and is a severe impact to server performance. Below are the profiler readings.
```
                                                         Profile results (total time)
Proc Name                                                                                      Self CPU    Total CPU    Real Time     Overtime        Calls
------------------------------------------------------------------------------------------    ---------    ---------    ---------    ---------    ---------
/mob/proc/create_chat_message                                                                     0.568       23.065       23.102        0.281       393209
/datum/chatmessage/proc/generate_image                                                           40.810       60.854   105850.133        7.320       343809
/datum/chatmessage/proc/end_of_life                                                              11.669       29.924       28.463        0.000       367658
/datum/chatmessage/New                                                                            0.857       22.503       22.508        0.434       343809
/datum/chatmessage/Destroy                                                                        6.171        9.027        8.582        0.000       343278
/datum/chatmessage/proc/colorize_string                                                           0.011        0.032        0.023        0.001          834
/datum/controller/subsystem/timer/fire                                                            4.494       85.376       81.925        0.001       109593
```
`Real Time` can be ignored as that includes async delay time, which includes messsage fades. Generating the images themselves causes 7 seconds of overtime, and 60 full seconds of CPU usage in a round. This is on the line of old tcomms in terms of excess CPU.

TLDR: Looks cool and is good from a design point, but good lord does it GC awfully and it puts the server on its knees. 

I may look into making this implementation better, but its not a top priority, and plus were out of pref toggles until #13936 is merged, so this is effectively on hold.

**Comment by Tarhalindur** [comment]
@procdrone

The problem isn't that I think it's ugly. The problem is that it does *literally nothing* to improve the gameplay experience.
Like, how long is runechat awe going to last? A week? Maybe a month?
It doesn't make your life more convenient, it doesn't add any other tangible improvements, and it's a performance hog.

**Comment by procdrone** [comment]
> @procdrone
> 
> The problem isn't that I think it's ugly. The problem is that it does _literally nothing_ to improve the gameplay experience.
> Like, how long is runechat awe going to last? A week? Maybe a month?
> It doesn't make your life more convenient, it doesn't add any other tangible improvements, and it's a performance hog.

"Thats like, your opinion". For me its an convenience. an improvement. is nice. it helps. 

Just because you don't see a reason doesn't mean there isn't any. Also Runechat is a feature on almost all ss13 servers since few months now. and it stayed. 

And I did admit to performance issues being valid. Now lets not draw this PR out in discussions on personal taste. The toggle option is for people that dont like it, like you.

**Comment by ZomgPonies** [comment]
@AffectedArc07 Curious to see what improvements you can make to it down the road.

**Comment by Tokorizo** [comment]
> @procdrone
> 
> The problem isn't that I think it's ugly. The problem is that it does _literally nothing_ to improve the gameplay experience.
> Like, how long is runechat awe going to last? A week? Maybe a month?
> It doesn't make your life more convenient, it doesn't add any other tangible improvements, and it's a performance hog.
It is a massive convenience and quality of life change, if you don't like it turn it off, performance-wise im sure there is something that can be done to improve it further down the line.

File: code/controllers/subsystem/afk.dm
Comment (AffectedArc07) [inline code/controllers/subsystem/afk.dm:82]: Why the hell is this in here?

**Review by Kyep — APPROVED** [review]
The final version needs:
- SQL support (persistent preference saving, I know its on your TO-DO list, it just can't be in here right now for TM reasons)
- Some performance stuff you're already working on.

Assuming we get both of those, I think we should merge this. While not everyone will use it, those who don't like it can just turn it off, and so long as it doesn't seriously damage performance for the rest of us, the degree to which people want this makes it worth having.

**Comment by AffectedArc07** [comment]
The OP on the PR has been updated and re-written. **Please, re read it**

File: code/game/atoms.dm
Comment (AffectedArc07) [inline code/game/atoms.dm:51]: 2

File: code/datums/chatmessage.dm
Comment (AffectedArc07) [inline code/datums/chatmessage.dm:261]: TODO: Document

File: code/modules/client/preference/preferences_toggles.dm
Comment (AffectedArc07) [inline code/modules/client/preference/preferences_toggles.dm:315]: Todo: Remove TM comment, use proper SQL support

File: code/modules/client/preference/preferences.dm
Comment (AffectedArc07) [inline code/modules/client/preference/preferences.dm:210]: TODO: SQL

File: code/modules/mob/living/autohiss.dm
Comment (AffectedArc07) [inline code/modules/mob/living/autohiss.dm:64]: TODO: Remove dirty diff

File: code/modules/mob/living/carbon/human/human.dm
Comment (AffectedArc07) [inline code/modules/mob/living/carbon/human/human.dm:1943]: TODO: Document

File: interface/skin.dmf
Comment (AffectedArc07) [inline interface/skin.dmf:1]: TODO: Unfuck line endings

File: paradise.dme
Comment (AffectedArc07) [inline paradise.dme:1]: TODO: Unfuck line endings

**Comment by Carthusia** [comment]
Something i noticed during the test, was that it said (says) before everything anyone said in the bubble chat.

It seemed kind of redundant to have: says Hi there:, floating over Bobs head, when its obvious he is saying it, or, and i never saw this during the test, is this because it also shows emotes, and *me's?

**Comment by ZomgPonies** [comment]
Yeah seeing the *bubble bubble bubble* in blue or whatever is an important indicator that things are being said and you can hear sound coming from the player, even if you can't understand the message itself.

**Comment by Varghh** [comment]
I think Runechat is amazing but i get the problem we have with our massive population. The only idea i get from the top of my head is to only use it for "say" speech, take any kind of radio chating out of runechat to diminish the amount of messages it will create. i feel it fits more for situation where people is talking next to you and not for every single radio message.

**Comment by AffectedArc07** [comment]
> Something i noticed during the test, was that it said (says) before everything anyone said in the bubble chat.
> 
> It seemed kind of redundant to have: says Hi there:, floating over Bobs head, when its obvious he is saying it, or, and i never saw this during the test, is this because it also shows emotes, and *me's?
> 
> Edit: This caught my eye:
> 
> "You will not see a message in a specified language if you do not have said language"
> 
> Does this mean you wont see it at all? So say: 4 vox are at the bar all speaking vox-pigeon and your a human, you won't see any text above their head in the usual weird fakey language you don't understand kind of text like normal? Will it just be a ?, or blank? The current description listed makes it seem like you won't see anything at all to indicate the player is speaking something you don't understand.

Youll see the gibberish the exact same way you would see it in chat. Emotes are not shown. The dupe say bug has been fixed.



> I think Runechat is amazing but i get the problem we have with our massive population. The only idea i get from the top of my head is to only use it for "say" speech, take any kind of radio chating out of runechat to diminish the amount of messages it will create. i feel it fits more for situation where people is talking next to you and not for every single radio message.

If performance requires this, I can easily gut the whole radio idea, as thats probably contributing to the massive amount of messages being created, and also not really needed.

**Comment by ZomgPonies** [comment]
@AffectedArc07 - Was there a chat window overhaul planned where you could like...seperate different channels into different tabs? I could totally see myself putting "say" stuff into it's own seperate tab and keeping all the rest into my 'main' chat window, relying on runechat for the say stuff 95% of the time.

**Comment by AffectedArc07** [comment]
@ZomgPonies see https://github.com/ParadiseSS13/Paradise/pull/13191

**Comment by Varghh** [comment]
> > I think Runechat is amazing but i get the problem we have with our massive population. The only idea i get from the top of my head is to only use it for "say" speech, take any kind of radio chating out of runechat to diminish the amount of messages it will create. i feel it fits more for situation where people is talking next to you and not for every single radio message.
> 
> If performance requires this, I can easily gut the whole radio idea, as thats probably contributing to the massive amount of messages being created, and also not really needed.

Runechat could just land a pop up with a speaker of the color of the channel that person is using, and just leave the floating text for person talking with "say" in your viewrange.

Let's say...
;HELP MAINTS -> This will only show up a green speaker on top of the poor bastard
:s Can i kill the clown already? -> Red Speaker on top of the officer head
Say: Captain you really need to stop doing meth -> Pop ups on top of the head of the worried blueshield

**Comment by AffectedArc07** [comment]
> > > I think Runechat is amazing but i get the problem we have with our massive population. The only idea i get from the top of my head is to only use it for "say" speech, take any kind of radio chating out of runechat to diminish the amount of messages it will create. i feel it fits more for situation where people is talking next to you and not for every single radio message.
> > 
> > 
> > If performance requires this, I can easily gut the whole radio idea, as thats probably contributing to the massive amount of messages being created, and also not really needed.
> 
> Runechat could just land a pop up with a speaker of the color of the channel that person is using, and just leave the floating text for person talking with "say" in your viewrange.
> 
> Let's say...
> ;HELP MAINTS -> This will only show up a green speaker on top of the poor bastard
> :s Can i kill the clown already? -> Red Speaker on top of the officer head
> Say: Captain you really need to stop doing meth -> Pop ups on top of the head of the worried blueshield

Not how this works anyway. You only see the message above someones head if you have that radio channel, because it works off of `hear_radio`. The blueshield can speak down command all they want, and even if youre adjacent, you wont see the speaker icon. You only see the speaker icon if you have that radio channel.

**Comment by Spacemanspark** [comment]
Emotes using the ! key don't work at all anymore with runechat.
While I was already on the fence about this feature, I'm certainly not willing to consider it until that's fixed.

Also arc when are we getting runecha-

**Comment by AffectedArc07** [comment]
> Emotes using the ! key don't work at all anymore with runechat.
> While I was already on the fence about this feature, I'm certainly not willing to consider it until that's fixed.

Will look into fixing this

> Also arc when are we getting runecha-

Will look into stabbing spark

**Comment by AffectedArc07** [comment]
Following a testmerge, I discovered that a lot more people have dark hair ingame than I expected. Colour now has a lower bound of how dark it can be.
image
image

**Comment by AffectedArc07** [comment]
Did a testmerge with about ~80 people average throughout

```
                                                         Profile results (total time)
Proc Name                                                                                       Self CPU    Total CPU    Real Time     Overtime        Calls
-------------------------------------------------------------------------------------------    ---------    ---------    ---------    ---------    ---------
/datum/chatmessage/proc/generate_image                                                             5.800        9.258    43549.434        0.531       122780
/datum/chatmessage/New                                                                             0.311        6.809        6.860        0.050       122780
/datum/chatmessage/proc/end_of_life                                                                3.766        4.500        4.496        0.000       122608
/datum/chatmessage/Destroy                                                                         2.292        3.655        3.684        0.000       122608
/datum/chatmessage/proc/enter_subsystem                                                            0.907        0.915        0.938        0.010       262986
/datum/chatmessage/proc/leave_subsystem                                                            0.308        0.311        0.345        0.000       122608
/datum/chatmessage/proc/colorize_string                                                            0.033        0.057        0.061        0.003         1854
/datum/controller/subsystem/runechat/fire                                                          1.226       11.555       11.601        0.000       159090
```

```
[2020-08-27T15:16:46] Path: /datum/timedevent
[2020-08-27T15:16:46] 	qdel() Count: 557032
[2020-08-27T15:16:46] 	Destroy() Cost: 5007ms
[2020-08-27T15:16:46] Path: /datum/chatmessage
[2020-08-27T15:16:46] 	qdel() Count: 125647
[2020-08-27T15:16:46] 	Destroy() Cost: 3957ms
```

The GC queue was much lower, but this was also lower population. Performance seems "manageable" though.

**Comment by AffectedArc07** [comment]
People inside lockers will no longer have their runechats displayed on the locker itself, same with chameleon projectors, ventcrawlers and more.

**Comment by TheYeetster1** [comment]
I created a account just to make this comment. During the test merge I was playing security officer, Processing people was much easier as I could quickly Identify who was talking, And even though there was 5 people in processing all talking, it was clear what was going on. This PR helps RP and I strongly recommend its implementation if possible.

**Comment by AffectedArc07** [comment]
Tweaks in this update:
- Cleaned up some excess code in places
- Made `get_runechat_color` defined on the `/atom` level so that anything can override this for a specific subtype generating colours (example: overriding on consoles to make them speak in their screen colour properly)
- Fixed italic text and cleaned up some spans from the skin
- Tweaked the radio code more to remove some excess stuff
- Cleaned up some proc args
- Messed up the line endings on `skin.dmf` again

**Comment by AffectedArc07** [comment]
TODO:
- Deconflict
- Fix holopad talk

**Comment by AffectedArc07** [comment]
Holopads have been fixed.

Before:
image

After:
image

*Note the text wont always be grey, these two test characters just had the default hair colour*

**Comment by AffectedArc07** [comment]
```
/datum/chatmessage
- qdel() Count: 64215
- Total Destroy() Cost: 2184ms
- Average Destroy() Cost: 0.034ms
```
^ Latest TM qdel result

```
                                                       Profile results (total time)
Proc Name                                                                                   Self CPU    Total CPU    Real Time     Overtime        Calls
---------------------------------------------------------------------------------------    ---------    ---------    ---------    ---------    ---------
/datum/controller/subsystem/runechat/fire                                                      0.873        6.533        6.550        0.000       147555
/mob/living/carbon/human/get_runechat_color                                                    0.062        0.167        0.174        0.008        57507
/datum/species/proc/get_species_runechat_color                                                 0.060        0.100        0.104        0.009        56460
/client/verb/toggle_runechat                                                                   0.001        0.007        0.007        0.000           48
/atom/proc/get_runechat_color                                                                  0.004        0.004        0.004        0.001         6448
/datum/species/grey/get_species_runechat_color                                                 0.001        0.002        0.003        0.000         1047
/datum/controller/subsystem/runechat/stat_entry                                                0.000        0.003        0.003        0.000          190
/datum/controller/subsystem/runechat/fire                                                      0.873        6.533        6.550        0.000       147555
/mob/living/carbon/human/get_runechat_color                                                    0.062        0.167        0.174        0.008        57507
/datum/species/proc/get_species_runechat_color                                                 0.060        0.100        0.104        0.009        56460
/datum/chatmessage/proc/generate_image                                                         5.119        8.973    21589.590        0.408        63955
/datum/chatmessage/New                                                                         0.146        5.792        5.785        0.020        63955
/datum/chatmessage/proc/end_of_life                                                            1.996        2.419        2.428        0.000        63901
/datum/chatmessage/Destroy                                                                     1.316        2.067        2.070        0.000        63904
/datum/chatmessage/proc/sanitize_color                                                         0.521        2.013        2.019        0.090        63955
/datum/chatmessage/proc/enter_subsystem                                                        0.607        0.614        0.634        0.009       138934
/datum/chatmessage/proc/leave_subsystem                                                        0.194        0.196        0.198        0.000        63904
/datum/chatmessage/proc/colorize_string                                                        0.006        0.023        0.023        0.002          507
```
^ Latest TM profile

I think this *should* be fine, though ill let other maints make the decision

**Comment by Spacemanspark** [comment]
Bad, start from scrap and do it all over again

**Comment by AffectedArc07** [comment]
Ghosts not seeing runechats has now been deemed intentional as a ghost could enable seeing all chat messages in the world, and potentially bog down the server from the amount of messages. 

This PR will stay in the draft state until https://github.com/ParadiseSS13/Paradise/pull/13936 is merged

**Comment by AffectedArc07** [comment]
SQL Query to enable runechat for all pre-existing players
```sql
UPDATE `player` SET `toggles_2` = `toggles_2` + 64
```

File: code/modules/mob/living/carbon/human/species/grey.dm
Comment (Edan52) [inline code/modules/mob/living/carbon/human/species/grey.dm:92]: Need to account for drask and kidan too.

Comment (AffectedArc07) [reply code/modules/mob/living/carbon/human/species/grey.dm:92]: image
image

**Comment by Spacemanspark** [comment]
Yo arc when we getting that iPhone imessage chat system

**Comment by AffectedArc07** [comment]
@Spacemanspark 
image

**Comment by Spacemanspark** [comment]
Hot

**Pull request merged**
